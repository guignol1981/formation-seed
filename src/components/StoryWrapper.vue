<template>
    <div class="m-u--app-body">
        <slot></slot>
    </div>
</template>

<script lang="ts">
    import Vue from 'vue';
    import Component from 'vue-class-component';

    @Component
    export default class StoryWrapper extends Vue {}
</script>

<style lang="scss">
    $m-font-path: '~@ulaval/modul-components/dist/assets/fonts/'; //this is required.
    @import '~@ulaval/modul-components/dist/styles/main.scss';
</style>
