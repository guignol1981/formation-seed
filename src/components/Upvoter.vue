<template>
    <div>
        <m-icon-button
            iconName="m-svg__arrow--up"
            buttonSize="32px"
            iconSize="20px"
            :skin="ownerHasUpvoted ? 'primary' : 'light'"
            @click="onUpvoteClicked()"
        ></m-icon-button>
        {{ votable.upvotes.length }}
        <m-icon-button
            iconName="m-svg__arrow--down"
            buttonSize="32px"
            iconSize="20px"
            :skin="ownerHasDownvoted ? 'primary' : 'light'"
            @click="onDownvoteClicked()"
        ></m-icon-button>
        {{ votable.downvotes.length }}
    </div>
</template>

<script lang="ts">
    export interface Votable {
        upvotes: string[];
        downvotes: string[];
    }

    import Vue from 'vue';
    import Component from 'vue-class-component';
    import { Prop, Emit, Watch } from 'vue-property-decorator';
    const _ = require('lodash');

    @Component
    export default class VueVoter extends Vue {
        @Prop({ required: true })
        public votable!: Votable;

        @Prop({ required: true })
        public ownerId!: string;

        public get ownerHasUpvoted(): boolean {
            return this.votable.upvotes.includes(this.ownerId);
        }

        public get ownerHasDownvoted(): boolean {
            return this.votable.downvotes.includes(this.ownerId);
        }

        public onUpvoteClicked(): void {
            if (this.ownerHasUpvoted) {
                this.votable.upvotes = this.votable.upvotes.filter(
                    u => u !== this.ownerId
                );
                return;
            }

            this.votable.upvotes.push(this.ownerId);
            this.votable.downvotes = this.votable.downvotes.filter(
                d => d !== this.ownerId
            );
        }

        public onDownvoteClicked(): void {
            if (this.ownerHasDownvoted) {
                this.votable.downvotes = this.votable.downvotes.filter(
                    d => d !== this.ownerId
                );
                return;
            }

            this.votable.downvotes.push(this.ownerId);
            this.votable.upvotes = this.votable.upvotes.filter(
                u => u !== this.ownerId
            );
        }
    }
</script>

<style lang="scss" scoped></style>
