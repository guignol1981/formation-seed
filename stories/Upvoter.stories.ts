import { Component } from 'vue';
import Upvoter from '../src/components/Upvoter';
import StoryWrapper from '../src/components/StoryWrapper';
import StoryRouter from 'storybook-vue-router';
import { action } from '@storybook/addon-actions';
import { withKnobs } from '@storybook/addon-knobs';

export default {
    title: 'Upvoter',
    component: Upvoter,
    decorators: [StoryRouter(), withKnobs]
};

export const Default: () => Component = (): Component => ({
    components: { Upvoter, StoryWrapper },
    data: () => ({
        votable: {
            upvotes: ['3', '2', '3'],
            downvotes: ['1']
        },
        ownerId: '1'
    }),
    template: `
        <StoryWrapper>
                <Upvoter 
                    :ownerId="ownerId"
                    :votable="votable"
                    :ownerId="ownerId"
                >
                </Upvoter>
        </StoryWrapper>
    `,
    methods: {
        onEditClicked: action('editClicked')
    }
});
