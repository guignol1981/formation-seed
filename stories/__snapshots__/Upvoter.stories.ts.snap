// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots Upvoter Default 1`] = `
<div
  class="m-u--app-body"
>
  <div>
    <button
      class="m-icon-button m--is-skin-light m--has-ripple"
      data-v-4b160bdb=""
      style="width: 32px; height: 32px;"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="m-icon m-icon-button__icon"
        data-v-3d21e689=""
        data-v-4b160bdb=""
        height="20px"
        width="20px"
      >
        <!---->
         
        <use
          aria-hidden="true"
          class=""
          data-v-3d21e689=""
          xlink:href="#m-svg__arrow--up"
        />
      </svg>
       
      <!---->
    </button>
    
    3
    
    <button
      class="m-icon-button m--is-skin-primary m--has-ripple"
      data-v-4b160bdb=""
      style="width: 32px; height: 32px;"
      type="button"
    >
      <svg
        aria-hidden="true"
        class="m-icon m-icon-button__icon"
        data-v-3d21e689=""
        data-v-4b160bdb=""
        height="20px"
        width="20px"
      >
        <!---->
         
        <use
          aria-hidden="true"
          class=""
          data-v-3d21e689=""
          xlink:href="#m-svg__arrow--down"
        />
      </svg>
       
      <!---->
    </button>
    
    1

  </div>
</div>
`;
